segmentstore/server/src/test/java/io/pravega/segmentstore/server/DataRecoveryTestUtils.java

```diff
@@ -0,0 +1,188 @@
+            it.forEachRemaining(k -> segmentsInMD.addAll(k.getEntries()), EXECUTOR_SERVICE).join();
+
+            for (SegmentProperties segment : segments) {
+                long len = segment.getLength();
+                boolean isSealed = segment.isSealed();
+                String segmentName = segment.getName();
+
+                /*
+                    1. segment exists in both metadata and storage, re-create it
+                    2. segment only in metadata, delete
+                    3. segment only in storage, re-create it
+                 */
+                segmentsInMD.remove(TableKey.unversioned(getTableKey(segmentName)));
+                container.getStreamSegmentInfo(segment.getName(), TIMEOUT)
+                        .thenAccept(e -> {
+                            container.createStreamSegment(segmentName, len, isSealed)
+                                    .exceptionally(ex -> {
+                                        log.error("Exception occurred while creating segment", ex);
+                                        return null;
+                                    }).join();
+                        })
+                        .exceptionally(e -> {
+                            log.error("Got an exception on getStreamSegmentInfo", e);
+                            if (Exceptions.unwrap(e) instanceof StreamSegmentNotExistsException) {
+                                container.createStreamSegment(segmentName, len, isSealed)
+                                        .exceptionally(ex -> {
+                                            log.error("Exception occurred while creating segment", ex);
+                                            return null;
+                                        }).join();
+                            }
+                            return null;
+                        }).join();
+            }
+            for (TableKey k : segmentsInMD) {
+                String segmentName = k.getKey().toString();
+                log.info("Deleting segment : {} as it is not in storage", segmentName);
+                try {
+                    container.deleteStreamSegment(segmentName, TIMEOUT).join();
+                } catch (Throwable e) {
+                    log.error("Error while deleting the segment = {}", segmentName);
+                }
+            }
+            log.info("Recovery done for container = {}", containerId);
+        }
+    }
+
+    public static ArrayView getTableKey(String segmentName) {
+        return new ByteArraySegment(segmentName.getBytes(Charsets.UTF_8));
+    }
+
+    /**
+     * Deletes container-metadata segment and attribute index segment for the given container Id.
+     * @param tier2         Long term storage to delete the segments from.
+     * @param containerId   Id of the container for which the segments has to be deleted.
+     */
+    public static void deleteContainerMetadataSegments(Storage tier2, int containerId) {
```

Comment:
Why is this static and public?

Reply:
And what's `tier2`?

Reply:
Made it private. But static, because static method `recoverAllSegments` uses it.